pkg/clusteragent/externalmetrics/autoscaler_watcher.go

```diff
@@ -0,0 +1,231 @@
+		_, active := datadogMetricReferences[datadogMetric.ID]
+		if active != datadogMetric.Active {
+			log.Debugf("Updating active status for: %s to: %t", datadogMetric.ID, active)
+
+			if currentDatadogMetric := w.store.LockRead(datadogMetric.ID, false); currentDatadogMetric != nil {
+				currentDatadogMetric.UpdateTime = time.Now().UTC()
+				currentDatadogMetric.Active = active
+				// If we move from Active to Inactive, we discard current valid state to avoid using unrefreshed metrics upon re-activation
+				if !currentDatadogMetric.Active {
+					currentDatadogMetric.Valid = false
+				}
+
+				w.store.UnlockSet(currentDatadogMetric.ID, *currentDatadogMetric, autoscalerWatcherStoreID)
+			}
+		}
+
+		// Delete autogen DatadogMetrics that haven't been updated for some time
+		if !active && datadogMetric.Autogen && !datadogMetric.HasBeenUpdatedFor(w.autogenExpirationPeriod) {
+			log.Infof("Flagging old autogen DatadogMetric: %s for deletion - last update: %v", datadogMetric.ID, datadogMetric.UpdateTime)
+			if currentDatadogMetric := w.store.LockRead(datadogMetric.ID, false); currentDatadogMetric != nil {
+				currentDatadogMetric.Deleted = true
+				w.store.UnlockSet(currentDatadogMetric.ID, *currentDatadogMetric, autoscalerWatcherStoreID)
+			}
+		}
+
+		// We clean reference map to keep only
+		delete(datadogMetricReferences, datadogMetric.ID)
+	}
+
+	// In `datadogMetricReferences` we now only have existing references that we should be able to handle, creating them
+	for datadogMetricID, externalMetric := range datadogMetricReferences {
+		autogenQuery := buildDatadogQueryForExternalMetric(externalMetric.metricName, externalMetric.metricLabels)
+		autogenDatadogMetric := model.NewDatadogMetricInternalFromExternalMetric(datadogMetricID, autogenQuery, externalMetric.metricName)
+		log.Infof("Creating DatadogMetric: %s for ExternalMetric: %s, Query: %s", datadogMetricID, externalMetric.metricName, autogenQuery)
+		w.store.Set(datadogMetricID, autogenDatadogMetric, autoscalerWatcherStoreID)
+	}
+}
+
+func (w *AutoscalerWatcher) getAutoscalerReferences() (map[string]*externalMetric, error) {
+	datadogMetricReferences := make(map[string]*externalMetric, w.store.Count())
+
+	// Helper func to avoid some copy paste between HPA and WPA
+	addAutogenReference := func(metricName string, labels map[string]string) {
+		datadogMetricName := getAutogenDatadogMetricNameFromLabels(metricName, labels)
+		datadogMetricID := w.autogenNamespace + kubernetesNamespaceSep + datadogMetricName
+		datadogMetricReferences[datadogMetricID] = &externalMetric{
+			metricName:        metricName,
+			metricLabels:      labels,
+			datadogMetricName: datadogMetricName,
+		}
+	}
+
+	if w.autoscalerLister != nil {
+		hpaList, err := w.autoscalerLister.HorizontalPodAutoscalers(metav1.NamespaceAll).List(labels.Everything())
+		if err != nil {
+			return nil, fmt.Errorf("Could not list HPAs (to update DatadogMetric active status): %v", err)
```

Comment:
But if the DM is associated to a WPA and for whatever reason the lister for HPA fails, we do not process the WPA [l:205](https://github.com/DataDog/datadog-agent/pull/5384/files/565a78c52098f2b6aab35f733003f3b23f2fde9b#diff-c34c8104d441fefbe8af5ba433cbb98cR205).

Reply:
Yes that's exactly what we want. If any of the lister fails and we continue processing (so processing only HPA or only WPA), we cannot know later on if the references are missing because a lister failed or because DDM reference has been removed from HPA/WPA.

Reply:
Understood, and I can see that we don't return if we get a HPA but we fail on listing WPAs
we don't return `datadogMetricReferences`.
One thing though it seems to me that the DM has to be use in conjuction with the WPA controller. I believe it should work if users are just relying on HPAs